File: JKIAPHelp/ActivityIndicator/JKIAPActivityIndicator.m
//
//  JKIAPActivityIndicatorView.m
//  JKIAPHelp
//
//  Created by kane on 2018/8/17.
//  Copyright © 2018年 kane. All rights reserved.
//

#import "JKIAPActivityIndicator.h"


@interface   JKIAPActivityIndicator()
{
    UIWindow *_window;
}
@property (nonatomic, strong) UIActivityIndicatorView *actIndicatorView;
  @property (nonatomic, strong)  UIVisualEffectView *activitybackView;
   @property (nonatomic, strong) UILabel *label;
@property (nonatomic, strong) UIView *backView;
@end

@implementation JKIAPActivityIndicator

- (instancetype)init
{
    self = [super init];
    if (self) {
        [self creatViews];
    }
    return self;
    
}
/**
 * 活动指示器弹出框开始
 */
- (void)showActivityWithMessage:(NSString *)msg{
   
    
         _window.hidden = NO;
            if (msg) {
                self.label.text = msg;
                  }
           if (self.actIndicatorView.isAnimating == YES) {
               return ;
           }
       
           self.backView.alpha = 0;
           [self layoutViews];
           [self.actIndicatorView startAnimating];
           [UIView animateWithDuration:0.5 animations:^{
               self.backView.alpha = 1;
           }];
          
      
}

/**
 * 活动指示器弹出框结束
 */
- (void)stop{
 
        [self.actIndicatorView stopAnimating];
        [UIView animateWithDuration:0.2 animations:^{
            self.backView.alpha = 0;
            [self.backView removeFromSuperview];
            self->_window.hidden = YES;
        } completion:^(BOOL finished) {           
        }];
   
}

- (void)creatViews{
    if (!_window) {
        
        _window = [UIWindow new];
        _window.backgroundColor = UIColor.clearColor;
        _backView = [UIView new];
        
        _backView.userInteractionEnabled = YES;
        _backView.backgroundColor = [UIColor clearColor];
        _actIndicatorView = [[UIActivityIndicatorView alloc] initWithActivityIndicatorStyle:UIActivityIndicatorViewStyleWhiteLarge];
        
        
        UIBlurEffect *blur = [UIBlurEffect effectWithStyle:UIBlurEffectStyleLight];
        UIVisualEffectView * visualEffectView = [[UIVisualEffectView alloc] initWithEffect:blur];
        visualEffectView.alpha = 1;
        _actIndicatorView.color = [UIColor blackColor];
        _activitybackView = visualEffectView;
        _activitybackView.layer.cornerRadius = 5;
        _label = [UILabel new];
        _label.textAlignment = NSTextAlignmentCenter;
        _label.font = [UIFont systemFontOfSize:14];
        [_backView setTranslatesAutoresizingMaskIntoConstraints:NO];
        [_activitybackView setTranslatesAutoresizingMaskIntoConstraints:NO];
        [_label setTranslatesAutoresizingMaskIntoConstraints:NO];
        [_actIndicatorView setTranslatesAutoresizingMaskIntoConstraints:NO];
    }
  
}

- (void)layoutViews{
    
    UIWindow *keyWindow = _window;
    if (![keyWindow.subviews containsObject:_backView]) {

        [keyWindow addSubview:_backView];
         NSLayoutConstraint *constrant11 = [NSLayoutConstraint constraintWithItem:_backView
                                                                        attribute:NSLayoutAttributeTop relatedBy:NSLayoutRelationEqual toItem:keyWindow attribute:NSLayoutAttributeTop multiplier:1.0 constant:0];
          NSLayoutConstraint *constrant12 = [NSLayoutConstraint constraintWithItem:_backView
                                                                         attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:keyWindow attribute:NSLayoutAttributeLeft multiplier:1.0 constant:0];
          NSLayoutConstraint *constrant13 = [NSLayoutConstraint constraintWithItem:_backView
                                                                         attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:keyWindow attribute:NSLayoutAttributeRight multiplier:1.0 constant:0];
          NSLayoutConstraint *constrant14 = [NSLayoutConstraint     constraintWithItem:_backView
                                                    attribute:NSLayoutAttributeBottom relatedBy:NSLayoutRelationEqual toItem:keyWindow attribute:NSLayoutAttributeBottom multiplier:1.0 constant:0];
        [keyWindow addConstraints:@[constrant11,constrant12,constrant13,constrant14]];
        
        
        
        [_backView addSubview:_activitybackView];
        NSLayoutConstraint *constrant21 = [NSLayoutConstraint constraintWithItem:_activitybackView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:_backView attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0];
        
        NSLayoutConstraint *constrant22 = [NSLayoutConstraint constraintWithItem:_activitybackView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:_backView attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:0];
        
       NSLayoutConstraint *constrant23 = [NSLayoutConstraint constraintWithItem:_activitybackView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationGreaterThanOrEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:95];
        
        NSLayoutConstraint *constrant24 = [NSLayoutConstraint constraintWithItem:_activitybackView attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationGreaterThanOrEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:130];
        [_backView addConstraints:@[constrant21,constrant22,constrant23,constrant24]];
        
        
          [_activitybackView.contentView addSubview:_label];
        NSLayoutConstraint *constrant31 = [NSLayoutConstraint constraintWithItem:_label attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationGreaterThanOrEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0 constant:25];
        
        NSLayoutConstraint *constrant32 = [NSLayoutConstraint constraintWithItem:_label attribute:NSLayoutAttributeLeft relatedBy:NSLayoutRelationEqual toItem:_activitybackView attribute:NSLayoutAttributeLeft multiplier:1.0 constant:5];
        NSLayoutConstraint *constrant33 = [NSLayoutConstraint constraintWithItem:_label attribute:NSLayoutAttributeRight relatedBy:NSLayoutRelationEqual toItem:_activitybackView attribute:NSLayoutAttributeRight multiplier:1.0 constant:-5];
        NSLayoutConstraint *constrant34 = [NSLayoutConstraint constraintWithItem:_label attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:_activitybackView attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:20];
        [_activitybackView addConstraints:@[constrant31,constrant32,constrant33,constrant34]];
        
        
        [_activitybackView.contentView addSubview:_actIndicatorView];
        NSLayoutConstraint *constrant41 = [NSLayoutConstraint constraintWithItem:_actIndicatorView attribute:NSLayoutAttributeCenterX relatedBy:NSLayoutRelationEqual toItem:_activitybackView attribute:NSLayoutAttributeCenterX multiplier:1.0 constant:0];
        
      NSLayoutConstraint *constrant42 = [NSLayoutConstraint constraintWithItem:_actIndicatorView attribute:NSLayoutAttributeCenterY relatedBy:NSLayoutRelationEqual toItem:_activitybackView attribute:NSLayoutAttributeCenterY multiplier:1.0 constant:-10];
        
        [_activitybackView addConstraints:@[constrant41,constrant42]];
        
    }
}



@end
    

File: IPADemo/JKIAPHelper/Other/JKIAPActivityIndicator.swift
//
//  JKIAPActivityIndicator.swift
//  JKIAPHelper
//
//  Created by Kane on 2019/8/7.
//  Copyright © 2019 kane. All rights reserved.
//

import Foundation



struct JKIAPActivityIndicator {
    
    let actIndicatorView :UIActivityIndicatorView = {
        let view = UIActivityIndicatorView(style: .whiteLarge)
       
        view.color = .black
       
        return view
    }()
    let activitybackView : UIVisualEffectView = {
        let blur = UIBlurEffect(style: .light)
        let  visualEffectView = UIVisualEffectView(effect: blur)
        visualEffectView.alpha = 1;
        visualEffectView.layer.cornerRadius = 5;
        return visualEffectView
    }()
    let backView : UIView = {
        let view = UIView()
        view.backgroundColor = .clear
        return view
    }()
    
  
        
    init() {
        
        actIndicatorView.translatesAutoresizingMaskIntoConstraints = false
        backView.translatesAutoresizingMaskIntoConstraints = false
        activitybackView.translatesAutoresizingMaskIntoConstraints = false
        backView.addSubview(activitybackView)
        activitybackView.addSubview(actIndicatorView)
       layoutViews()
    }
        /**
         * 活动指示器弹出框开始
         */
    
    func start() {
        DispatchQueue.main.async {
            if self.actIndicatorView.isAnimating{
                self.stop()
            }
            self.backView.alpha = 0
            self.actIndicatorView.startAnimating()
            self.backView.layoutIfNeeded()
            UIView.animate(withDuration: 0.5, animations: {
                self.backView.alpha = 1
            })
        }
    }
    
    
    /**
     * 活动指示器弹出框结束
     */
    func stop() {
        UIView.animate(withDuration: 0.5, animations: {
            self.backView.alpha = 0
        }) { (_) in
            self.backView.removeFromSuperview()
            self.actIndicatorView.stopAnimating()
        }
            
    }

  private  func layoutViews() {
        let keyWindow = UIApplication.shared.keyWindow
        keyWindow?.addSubview(backView)
        let constrant11 = NSLayoutConstraint(item: backView, attribute: .top, relatedBy: .equal, toItem: keyWindow, attribute: .top, multiplier: 1.0, constant: 0)
        let constrant12 = NSLayoutConstraint(item: backView, attribute: .left, relatedBy: .equal, toItem: keyWindow, attribute: .left, multiplier: 1.0, constant: 0)
        let constrant13 = NSLayoutConstraint(item: backView, attribute: .right, relatedBy: .equal, toItem: keyWindow, attribute: .right, multiplier: 1.0, constant: 0)
        let constrant14 = NSLayoutConstraint(item: backView, attribute: .bottom, relatedBy: .equal, toItem: keyWindow, attribute: .bottom, multiplier: 1.0, constant: 0)
        keyWindow?.addConstraints([constrant11, constrant12, constrant13, constrant14])
        
        
        
        let constrant21 = NSLayoutConstraint(item: activitybackView, attribute: .centerX, relatedBy: .equal, toItem: backView, attribute: .centerX, multiplier: 1.0, constant: 0)
        
        let constrant22 = NSLayoutConstraint(item: activitybackView, attribute: .centerY, relatedBy: .equal, toItem: backView, attribute: .centerY, multiplier: 1.0, constant: 0)
        
        let constrant23 = NSLayoutConstraint(item: activitybackView, attribute: .height, relatedBy: .greaterThanOrEqual, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: 95)
        
        let constrant24 = NSLayoutConstraint(item: activitybackView, attribute: .width, relatedBy: .greaterThanOrEqual, toItem: nil, attribute: .notAnAttribute, multiplier: 1.0, constant: 130)
        backView.addConstraints([constrant21, constrant22, constrant23, constrant24])

let constrant41 = NSLayoutConstraint(item: actIndicatorView, attribute: .centerX, relatedBy: .equal, toItem: activitybackView, attribute: .centerX, multiplier: 1.0, constant: 0)

let constrant42 = NSLayoutConstraint(item: actIndicatorView, attribute: .centerY, relatedBy: .equal, toItem: activitybackView, attribute: .centerY, multiplier: 1.0, constant: -10)

activitybackView.addConstraints([constrant41, constrant42])

    

    }
    
   

}
